app/models/user_subscription.rb

```diff
@@ -0,0 +1,12 @@
+class UserSubscription < ApplicationRecord
+  ALLOWED_TYPES = %w[Article].freeze
+
+  belongs_to :author, class_name: "User", foreign_key: :author_id, inverse_of: :user_subscriptions
+  belongs_to :subscriber, class_name: "User", foreign_key: :subscriber_id, inverse_of: :user_subscriptions
+  belongs_to :user_subscription_sourceable, polymorphic: true
```

Comment:
can you explain this?

Reply:
@rhymes sure! Do you mean how polymorphism works? Or the context of the relationships?

Reply:
This is part of the complexity of polymorphic and how you "auto-magically" setup polymorphic to work by telling this `UserSubscription` model that it's a part of that relationship. It's like saying, "the `UserSubscription` belongs to the `user_subscription_sourceable` entity". It's usually more straight forward when we keep the names the same (i.e. `:user_subscribable`). I don't know the technical term as far as grammar goes...but making the word "-able" is a convention for polymorphic associations. And when you look at the migration, you'll see:

```ruby
t.references :user_subscription_sourceable, polymorphic: true
```

From that, Rails will do its magic and create the `user_subscription_source_type` and `user_subscription_source_id` columns for you. `type` and `id` are what we need to create the object we're referencing. So that would look like: 
```
user_subscription_source_type: "Article"
user_subscription_source_id: 1
```
which is saying that the `source` of the `UserSubscription` is the `Article` object with `id` of `1`. It's convenient because Rails/ActiveRecord will return the actual `Article` record by simply calling `UserSubscription.first.user_subscription_sourceable`.

And the name `user_subscription_sourceable` is something we choose. In other words, we're saying this object is `sourceable` (can be a source) for a `UserSubscription`.

I think it's more clear with a generic example. Say you were to build some review website (like a Yelp clone). And say you had different types of things that could be reviewed (`Restaurant` and `Concert`). Those are 2 different object types that we want to associate with `Review`. So, in this example, we'd likely just call it `reviewable` which would make the columns `reviewable_type` and `reviewable_id`. We're saying both `Restaurants` and `Concerts` are `reviewable`.

In our example, we're saying a `UserSubscription`can be `sourced` from an `Article` or `Comment`.

Reply:
ahahahha sorry, I should have been more clear so to avoid having you spend so much time replying to what was essentially the following question: why do we differentiate sources? 

But you also replied to that just now, thank you and sorry again for having you explain the entire context! I just didn't understand why a person would be subscribed to either an article or a comment. But I was wrong, it's not subscribed to an article or a comment, it's tracking where the subscription came from!